File: edudesk-kotlin/src/main/kotlin/com/edudesk/services/AssignmentService.kt
package com.edudesk.services

import com.edudesk.models.Assignment
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.okhttp.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.serialization.json.Json
import kotlinx.serialization.Serializable

class AssignmentService {
    private val client = HttpClient(OkHttp) {
        install(ContentNegotiation) {
            json(Json { ignoreUnknownKeys = true })
        }
    }
    private val baseUrl = "http://localhost:8081"

    @Serializable
    private data class AddAssignmentRequest(val title: String, val desc: String, val deadline: String, val courseId: Int, val userId: Int)

    suspend fun getAssignmentsForUser(userId: Int): List<Assignment> {
        return try {
            val response = client.get("$baseUrl/assignments?userId=$userId")
            if (response.status == HttpStatusCode.OK) {
                response.body()
            } else {
                emptyList()
            }
        } catch (e: Exception) {
            println("Error fetching assignments: ${e.message}")
            emptyList()
        }
    }

    suspend fun addAssignment(title: String, desc: String, deadline: String, courseId: Int, userId: Int) {
        try {
            client.post("$baseUrl/assignments") {
                contentType(ContentType.Application.Json)
                setBody(AddAssignmentRequest(title, desc, deadline, courseId, userId))
            }
        } catch (e: Exception) {
            println("Error adding assignment: ${e.message}")
        }
    }

    suspend fun markDone(assignmentId: Int) {
        try {
            client.put("$baseUrl/assignments/$assignmentId/done")
        } catch (e: Exception) {
            println("Error marking assignment done: ${e.message}")
        }
    }

    suspend fun deleteAssignment(assignmentId: Int) {
        try {
            client.delete("$baseUrl/assignments/$assignmentId")
        } catch (e: Exception) {
            println("Error deleting assignment: ${e.message}")
        }
    }
}

File: old-golang-backup/services/assignment_service.go
package services

import (
	"time"
	"gorm.io/gorm"
	"edudesk/models"
)

type AssignmentService struct {
	DB *gorm.DB
}

func (s *AssignmentService) GetAssignments(userID uint) ([]models.Assignment, error) {
	var assignments []models.Assignment
	err := s.DB.Where("user_id = ?", userID).Find(&assignments).Error
	return assignments, err
}

func (s *AssignmentService) AddAssignment(a *models.Assignment) error {
	return s.DB.Create(a).Error
}

func (s *AssignmentService) UpdateAssignment(a *models.Assignment) error {
	return s.DB.Save(a).Error
}

func (s *AssignmentService) DeleteAssignment(id uint) error {
	return s.DB.Delete(&models.Assignment{}, id).Error
}

func (s *AssignmentService) MarkDone(id uint) error {
	return s.DB.Model(&models.Assignment{}).Where("id = ?", id).Update("is_done", true).Error
}

func (s *AssignmentService) UploadSubmission(assignmentID uint, filePath string) error {
	submission := models.Submission{
		AssignmentID: assignmentID,
		FilePath: filePath,
		SubmittedAt: time.Now(),
	}
	return s.DB.Create(&submission).Error
}

func (s *AssignmentService) GetDeadlineSoon(userID uint) ([]models.Assignment, error) {
	var assignments []models.Assignment
	soon := time.Now().Add(72 * time.Hour)
	err := s.DB.Where("user_id = ? AND deadline <= ? AND is_done = ?", userID, soon, false).Find(&assignments).Error
	return assignments, err
}
